File: src/main/java/org/jbb/big/BigFile.java
package org.jbb.big;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.ImmutableList;
import com.google.common.collect.ImmutableSet;

import java.io.Closeable;
import java.io.IOException;
import java.nio.ByteOrder;
import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.Set;

/**
 * A common superclass for Big files.
 *
 * @author Sergei Lebedev
 * @author Sergey Zherevchuk
 * @since 11/04/15
 */
abstract class BigFile<T> implements Closeable, AutoCloseable {

  @VisibleForTesting
  protected static class Header {

    static Header parse(final SeekableDataInput s, final int magic) throws IOException {
      s.guess(magic);

      final short version = s.readShort();
      final short zoomLevels = s.readShort();
      final long chromTreeOffset = s.readLong();
      final long unzoomedDataOffset = s.readLong(); // fullDataOffset in supplemental info
      final long unzoomedIndexOffset = s.readLong();
      final short fieldCount = s.readShort();
      final short definedFieldCount = s.readShort();
      final long asOffset = s.readLong();
      final long totalSummaryOffset = s.readLong();
      final int uncompressBufSize = s.readInt();
      final long reserved = s.readLong();  // currently 0.

      if (uncompressBufSize > 0) {
        throw new IllegalStateException("data compression is not supported");
      }

      if (reserved != 0) {
        throw new IllegalStateException("header extensions are not supported");
      }

      final List<ZoomLevel> zoomLevelList = new ArrayList<>();
      for (int i = 0; i < zoomLevels; i++) {
        zoomLevelList.add(new ZoomLevel(s.readInt(), s.readInt(),
                                        s.readLong(), s.readLong()));
      }

      final BPlusTree bpt = BPlusTree.read(s, chromTreeOffset);
      final RTreeIndex rti = RTreeIndex.read(s, unzoomedIndexOffset);
      return new Header(s.order(), version, zoomLevels,
                        unzoomedDataOffset, fieldCount, definedFieldCount,
                        asOffset, totalSummaryOffset, uncompressBufSize, zoomLevelList,
                        bpt, rti);
    }

    public final ByteOrder byteOrder;
    public final short version;
    public final short zoomLevels;
    public final long unzoomedDataOffset;  // FullDataOffset in table 5
    public final short fieldCount;
    public final short definedFieldCount;
    public final long asOffset;
    public final long totalSummaryOffset;
    public final int uncompressBufSize;
    public final List<ZoomLevel> zoomLevelList;
    public final BPlusTree bPlusTree;
    public final RTreeIndex rTree;

    protected Header(final ByteOrder byteOrder, final short version,
                     final short zoomLevels, final long unzoomedDataOffset,
                     final short fieldCount, final short definedFieldCount, final long asOffset,
                     final long totalSummaryOffset, final int uncompressBufSize,
                     final List<ZoomLevel> zoomLevelList,
                     final BPlusTree bPlusTree, final RTreeIndex rTree) {
      this.byteOrder = byteOrder;
      this.version = version;
      this.zoomLevels = zoomLevels;
      this.unzoomedDataOffset = unzoomedDataOffset;
      this.fieldCount = fieldCount;
      this.definedFieldCount = definedFieldCount;
      this.asOffset = asOffset;
      this.totalSummaryOffset = totalSummaryOffset;
      this.uncompressBufSize = uncompressBufSize;
      this.zoomLevelList = zoomLevelList;
      this.bPlusTree = bPlusTree;
      this.rTree = rTree;
    }
  }

  // XXX maybe we should make it a DataIO instead of separate
  // Input/Output classes?
  protected final SeekableDataInput handle;
  protected final Header header;

  protected BigFile(final Path path) throws IOException {
    this.handle = SeekableDataInput.of(path);
    this.header = BigFile.Header.parse(handle, getHeaderMagic());
  }

  public Set<String> chromosomes() throws IOException {
    final ImmutableSet.Builder<String> b = ImmutableSet.builder();
    header.bPlusTree.traverse(handle, bpl -> b.add(bpl.key));
    return b.build();
  }

  /**
   * Defaults {@code maxItems} to {@code 0}.
   */
  public List<T> query(final String chromName, final int startOffset, final int endOffset)
      throws IOException {
    return query(chromName, startOffset, endOffset, 0);
  }

  /**
   * Queries an R+-tree.
   *
   * @param chromName human-readable chromosome name, e.g. {@code "chr9"}.
   * @param startOffset 0-based start offset (inclusive).
   * @param endOffset 0-based end offset (exclusive), if 0 than the whole
   *                  chromosome is used.
   * @param maxItems upper bound on the number of items to return.
   * @return a list of intervals overlapping the query.
   * @throws IOException if the underlying {@link SeekableDataInput} does so.
   */
  public List<T> query(final String chromName, final int startOffset, final int endOffset,
                       final int maxItems)
      throws IOException {
    final Optional<BPlusLeaf> bpl = header.bPlusTree.find(handle, chromName);
    if (bpl.isPresent()) {
      final int chromIx = bpl.get().id;
      final RTreeInterval query = RTreeInterval.of(
          chromIx, startOffset, endOffset == 0 ? bpl.get().size : endOffset);
      return queryInternal(query, maxItems);
    } else {
      return ImmutableList.of();
    }
  }

  public abstract int getHeaderMagic();

  protected abstract List<T> queryInternal(RTreeInterval query, int maxItems)
      throws IOException;

  @Override
  public void close() throws IOException {
    handle.close();
  }
}

File: src/main/kotlin/org/jbb/big/BigFile.kt
package org.jbb.big

import com.google.common.collect.ImmutableList
import com.google.common.primitives.Ints
import com.google.common.primitives.Longs
import com.google.common.primitives.Shorts
import java.io.Closeable
import java.io.IOException
import java.nio.ByteOrder
import java.nio.file.Path
import kotlin.properties.Delegates

/**
 * A common superclass for Big files.
 */
abstract class BigFile<T> protected constructor(path: Path, magic: Int) :
        Closeable, AutoCloseable {

    val input: SeekableDataInput = SeekableDataInput.of(path)
    val header: Header = Header.read(input, magic)
    val zoomLevels: List<ZoomLevel> = (0 until header.zoomLevelCount.toInt()).asSequence()
            .map { ZoomLevel.read(input) }.toList()
    val bPlusTree: BPlusTree
    val rTree: RTreeIndex

    init {
        // Skip AutoSQL string if any.
        while (header.asOffset > 0 && input.readByte() != 0.toByte()) {}

        // Skip total summary block.
        if (header.totalSummaryOffset > 0) {
            Summary.read(input)
        }

        // Skip extended header. Ideally, we should issue a warning
        // if extensions are present.
        if (header.extendedHeaderOffset > 0) {
            with(input) {
                skipBytes(Shorts.BYTES)  // extensionSize.
                val extraIndexCount = readShort().toInt()
                skipBytes(Longs.BYTES)   // extraIndexListOffset.
                skipBytes(48)            // reserved.

                for (i in 0 until extraIndexCount) {
                    val type = readShort()
                    assert(type == 0.toShort())
                    val extraFieldCount = readShort()
                    skipBytes(Longs.BYTES)      // indexOffset.
                    skipBytes(extraFieldCount *
                              (Shorts.BYTES + // fieldId,
                               Shorts.BYTES))   // reserved.
                }
            }
        }

        bPlusTree = BPlusTree.read(input, header.chromTreeOffset)
        check(bPlusTree.header.order == header.order)
        rTree = RTreeIndex.read(input, header.unzoomedIndexOffset)
        check(rTree.header.order == header.order)
    }

    public val chromosomes: List<String> by Delegates.lazy {
        // XXX I wonder if sequential numbering is always the case for
        // Big files?
        with (bPlusTree) {
            val res = arrayOfNulls<String>(header.itemCount.toInt())
            for (leaf in traverse(input)) {
                res[leaf.id] = leaf.key
            }

            res.asSequence().filterNotNull().toArrayList()
        }
    }

    public val compressed: Boolean get() {
        // Compression was introduced in version 3 of the format. See
        // bbiFile.h in UCSC sources.
        return header.version >= 3 && header.uncompressBufSize > 0
    }

    /**
     * Queries an R+-tree.
     *
     * @param name human-readable chromosome name, e.g. `"chr9"`.
     * @param startOffset 0-based start offset (inclusive).
     * @param endOffset 0-based end offset (exclusive), if 0 than the whole
     *                  chromosome is used.
     * @return a list of intervals completely contained within the query.
     * @throws IOException if the underlying [SeekableDataInput] does so.
     */
    throws(IOException::class)
    public fun query(name: String, startOffset: Int, endOffset: Int): Sequence<T> {
        val res = bPlusTree.find(input, name)
        return if (res == null) {
            emptySequence()
        } else {
            val (_key, chromIx, size) = res
            val query = Interval.of(chromIx, startOffset,
                                    if (endOffset == 0) size else endOffset)
            rTree.findOverlappingBlocks(input, query)
                    .flatMap { queryInternal(it.dataOffset, it.dataSize, query) }
        }
    }

    throws(IOException::class)
    protected abstract fun queryInternal(dataOffset: Long, dataSize: Long,
                                         query: ChromosomeInterval): Sequence<T>

    throws(IOException::class)
    override fun close() = input.close()

    class Header(val order: ByteOrder, val version: Short, val zoomLevelCount: Short,
                 val chromTreeOffset: Long, val unzoomedDataOffset: Long,
                 val unzoomedIndexOffset: Long, val fieldCount: Short,
                 val definedFieldCount: Short, val asOffset: Long,
                 val totalSummaryOffset: Long, val uncompressBufSize: Int,
                 val extendedHeaderOffset: Long) {

        fun write(output: SeekableDataOutput, magic: Int) = with(output) {
            seek(0L)  // a header is always first.
            writeInt(magic)
            writeShort(version.toInt())
            writeShort(zoomLevelCount.toInt())
            writeLong(chromTreeOffset)
            writeLong(unzoomedDataOffset)
            writeLong(unzoomedIndexOffset)
            writeShort(fieldCount.toInt())
            writeShort(definedFieldCount.toInt())
            writeLong(asOffset)
            writeLong(totalSummaryOffset)
            writeInt(uncompressBufSize)
            writeLong(extendedHeaderOffset)
        }

        companion object {
            /** Number of bytes used for this header. */
            val BYTES = 64

            throws(IOException::class)
            fun read(input: SeekableDataInput, magic: Int): Header = with(input) {
                guess(magic)

                val version = readShort()
                val zoomLevelCount = readShort()
                val chromTreeOffset = readLong()
                val unzoomedDataOffset = readLong()
                val unzoomedIndexOffset = readLong()
                val fieldCount = readShort()
                val definedFieldCount = readShort()
                val asOffset = readLong()
                val totalSummaryOffset = readLong()
                val uncompressBufSize = readInt()
                val extendedHeaderOffset = readLong()
                return Header(order, version, zoomLevelCount, chromTreeOffset,
                              unzoomedDataOffset, unzoomedIndexOffset,
                              fieldCount, definedFieldCount, asOffset,
                              totalSummaryOffset, uncompressBufSize,
                              extendedHeaderOffset)
            }
        }
    }
}

data class ZoomLevel(public val reductionLevel: Int,
                     public val dataOffset: Long,
                     public val indexOffset: Long) {
    companion object {
        fun read(input: SeekableDataInput) = with(input) {
            val reductionLevel = readInt()
            val reserved = readInt()
            check(reserved == 0)
            val dataOffset = readLong()
            val indexOffset = readLong()
            ZoomLevel(reductionLevel, dataOffset, indexOffset)
        }
    }
}

data class Summary(public val basesCovered: Long,
                   public val minVal: Double,
                   public val maxVal: Double,
                   public val sumData: Double,
                   public val sumSquared: Double) {
    companion object {
        fun read(input: SeekableDataInput) = with(input) {
            val basesCovered = readLong()
            val minVal = readDouble()
            val maxVal = readDouble()
            val sumData = readDouble()
            val sumSquared = readDouble()
            Summary(basesCovered, minVal, maxVal, sumData, sumSquared)
        }
    }
}